src/TwigComponent/src/ComponentRenderer.php

```diff
@@ -20,19 +21,20 @@
  */
 final class ComponentRenderer
 {
-    private $twig;
+    private Environment $twig;
 
     public function __construct(Environment $twig)
     {
         $this->twig = $twig;
     }
 
-    public function render(ComponentInterface $component): string
+    public function render(object $component): string
     {
-        // TODO: Template attribute/annotation/interface to customize
         // TODO: Self-Rendering components?
-        $templateName = sprintf('components/%s.html.twig', $component::getComponentName());
+        if (!$attribute = TwigComponent::forClass($component::class)) {
+            throw new \InvalidArgumentException(sprintf('"%s" is not a Twig Component, did you forget to add the TwigComponent attribute?', $component::class));
```

Comment:
Doesn't this block users from specifying the component name and template directly in the tag (without using the attribute)?

I would go the route that the attribute is only used for auto-configuration and that the actual tag is used to find the name and template in all other places. That way you support both ways of configuring the component

Reply:
I think this is similar to what @wouterj was thinking as well. I suppose we could have some kind of component name => template map injected into the `ComponentRenderer` but the only way to get the component name from an instance is to check the attribute...

Reply:
I would pass the template to the render function actually. It makes sense to me to say: "render this component using this template". Theoretically you could tag multiple services with the same component class passing a different template (although in practice I don't think this will be common)

Reply:
That pushes the problem up to the twig extension (and, for live components, the event subscriber). Somehow, for `{{ component('my_component') }}` *my_component* needs to be mapped to a template. Or are you thinking the template should be passed to the component twig function?

Reply:
Hmm true. What about building a name => template map in the `TwigComponentPass` and setting the map as an argument to the `ComponentRuntime`? So it'd be something like this:

```php
public function __construct(ComponentFactory $componentFactory, ComponentRenderer $componentRenderer, array $templateMapping)
    {
        $this->componentFactory = $componentFactory;
        $this->componentRenderer = $componentRenderer;
        $this->templateMapping = $templateMapping;
    }

    public function render(string $name, array $props = []): string
    {
        return $this->componentRenderer->render(
            $this->componentFactory->create($name, $props),
            $this->templateMapping[$name]
        );
    }

```

Reply:
This would work in the `TwigComponent` package but would create a problem in the `LiveComponent` package. It needs to [pass an already created component instance to render](https://github.com/symfony/ux/blob/c118092650561e8e92cf87edf81448690b310da0/src/LiveComponent/src/EventListener/LiveComponentSubscriber.php#L234).

Reply:
The subscriber has access to the container so you could just get the ComponentRuntime instead of ComponentRenderer to render the template right?

Reply:
Sorry, I'm not following. The `ComponentRuntime` example above wouldn't work for the subscriber, the component is already created.

Reply:
Maybe we can talk on the symfony slack about it, might be a bit easier. I'm not available tonight, but I'll try to be online in the evenings later this week